File: evaluation/S0/python/Medium/1184-all-elements-in-two-binary-search-trees.py
# Source: https://leetcode.com/problems/all-elements-in-two-binary-search-trees/   |   Difficulty: Medium
#
# Problem Description:
# Given two binary search trees root1 and root2, return a list containing all the integers from both trees sorted in ascending order.
#
# Example:
# Input: root1 = [2,1,4], root2 = [1,0,3]
# Output: [0,1,1,2,3,4]
#
# Constraints:
# The number of nodes in each tree is in the range [0, 5000].
# 	-105 <= Node.val <= 105
#

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def getAllElements(self, root1, root2):
        """
        :type root1: Optional[TreeNode]
        :type root2: Optional[TreeNode]
        :rtype: List[int]
        """
        def inorder_traversal(root, elements):
            if root:
                inorder_traversal(root.left, elements)
                elements.append(root.val)
                inorder_traversal(root.right, elements)

        elements1 = []
        elements2 = []
        
        inorder_traversal(root1, elements1)
        inorder_traversal(root2, elements2)
        
        # Merge the two sorted lists
        merged_elements = []
        i, j = 0, 0
        
        while i < len(elements1) and j < len(elements2):
            if elements1[i] < elements2[j]:
                merged_elements.append(elements1[i])
                i += 1
            else:
                merged_elements.append(elements2[j])
                j += 1
        
        # Append remaining elements
        while i < len(elements1):
            merged_elements.append(elements1[i])
            i += 1
        
        while j < len(elements2):
            merged_elements.append(elements2[j])
            j += 1
        
        return merged_elements

File: evaluation/S0/javascript/Medium/1184-all-elements-in-two-binary-search-trees.js
// Source: https://leetcode.com/problems/all-elements-in-two-binary-search-trees/   |   Difficulty: Medium
//
// Problem Description:
// Given two binary search trees root1 and root2, return a list containing all the integers from both trees sorted in ascending order.
//
// Example:
// Input: root1 = [2,1,4], root2 = [1,0,3]
// Output: [0,1,1,2,3,4]
//
// Constraints:
// The number of nodes in each tree is in the range [0, 5000].
// 	-105 <= Node.val <= 105
//

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {number[]}
 */
var getAllElements = function(root1, root2) {
    const inorderTraversal = (node, result) => {
        if (!node) return;
        inorderTraversal(node.left, result);
        result.push(node.val);
        inorderTraversal(node.right, result);
    };
    const list1 = [];
    const list2 = [];
    inorderTraversal(root1, list1);
    inorderTraversal(root2, list2);
    const mergedList = [];
    let i = 0, j = 0;
    while (i < list1.length && j < list2.length) {
        if (list1[i] < list2[j]) {
            mergedList.push(list1[i]);
            i++;
        } else {
            mergedList.push(list2[j]);
            j++;
        }
    }
    while (i < list1.length) {
        mergedList.push(list1[i]);
        i++;
    }
    while (j < list2.length) {
        mergedList.push(list2[j]);
        j++;
    }
    return mergedList;
        
};
